### PR TITLE
feat(sns): Create test-only function that can be used to advance the target version directly (rather than through a proposal)

### PR DESCRIPTION
**Pull request by anchpop**
This will be useful for testing the new SNS upgrade flow, because we will be able to easily trigger the process at arbitrary times without the complexity of going through the (relatively more complex) proposal mechanism.

This PR also creates a `target_version` field in SNS Gov's proto.

### CODE REVIEW

File: rs/sns/governance/canister/canister.rs [hunk 1]
```diff
@@ -733,6 +734,14 @@ fn add_maturity_(request: AddMaturityRequest) -> AddMaturityResponse {
     governance_mut().add_maturity(request)
 }
 
+#[cfg(feature = "test")]
+#[export_name = "canister_update advance_target_version"]
+fn advance_target_version(request: AdvanceTargetVersionRequest) {
```
Comment (aterga) [inline rs/sns/governance/canister/canister.rs:739]: Why didn't this function end up in rs/sns/governance/canister/governance_test.did?

File: rs/sns/governance/canister/canister.rs [hunk 2]
```diff
@@ -733,6 +734,14 @@ fn add_maturity_(request: AddMaturityRequest) -> AddMaturityResponse {
...
+    over(candid_one, |request: AdvanceTargetVersionRequest| {
+        AdvanceTargetVersionResponse {}
```
Comment (aterga) [inline rs/sns/governance/canister/canister.rs:741]: Why not implement this function in the same PR?

**Review by aterga — APPROVED** [review]

File: rs/sns/governance/canister/canister.rs (hunk 1 above)
Comment (anchpop) [reply rs/sns/governance/canister/canister.rs:732]: Hmm, seems like it's not enforced by the candid type enforcement in C+

Comment (anchpop) [reply rs/sns/governance/canister/canister.rs:732]: No, the candid type enforecment seems to specifically reject it, not sure what's going on here

File: rs/sns/governance/canister/canister.rs (hunk 2 above)
Comment (anchpop) [reply rs/sns/governance/canister/canister.rs:734]: That is the plan, however the scope of this PR has unfortunately grown and I may have to start splitting it up before merging it

**Comment by anchpop** [comment]
The scope of this PR has expanded, to the point where it is now starting to implement Periodic Task A. I think that should be brought into a separate PR, with this one just keeping the AdvanceSnsVersion endpoint.

File: rs/sns/governance/canister/canister.rs (hunk 2 above)
Comment (aterga) [reply rs/sns/governance/canister/canister.rs:734]: Yes, let's keep this PR just for bumping the target in testing. 

Note that it should still only be possible to advance the target using new_target, not go back to an older version (I recommend implementing that logic in this PR as well).

File: rs/nervous_system/integration_tests/src/pocket_ic_helpers.rs [hunk 3]
```diff
@@ -1651,6 +1654,20 @@ pub mod sns {
...
+            sns_governance_canister_id: PrincipalId,
+            target_version: Version,
+        ) -> AdvanceTargetVersionResponse {
+            let payload = AdvanceTargetVersionRequest {
+                target_version: Some(target_version),
+            };
+            pocket_ic
+                .call(sns_governance_canister_id, payload)
+                .await
+                .unwrap()
```
Comment (anchpop) [inline rs/nervous_system/integration_tests/src/pocket_ic_helpers.rs:1669]: I implemented this with the CallCanisters trait as it's less verbose and does the same thing. However, for consistency, I'd be happy to change it to the way we typically call canisters in the pocketic integration tests.

**Review by aterga — APPROVED** [review]

**Pull request merged**